File: app/personnel_intake/domain/training_summary.py
"""Shared training summary calculations for intake draft and PPR training records."""
from __future__ import annotations

import calendar
from dataclasses import dataclass
from datetime import date
from decimal import Decimal, InvalidOperation
from typing import Any, Mapping, Sequence

TRAINING_WINDOW_YEARS = 5
CERTIFICATE_VALIDITY_YEARS = 5
CERTIFICATE_EXPIRY_LOOKAHEAD_MONTHS = 6
TRAINING_DOCUMENT_TYPE_CERTIFICATE = "certificate"


@dataclass(frozen=True, slots=True)
class TrainingSummaryRecord:
    title: str | None
    completed_at: date | None
    hours: Decimal | None
    document_type: str | None = None
    lifecycle_status: str | None = "ACTIVE"


@dataclass(frozen=True, slots=True)
class ExpiringCertificateSummary:
    title: str
    expires_at: date
    days_remaining: int


def training_window_start(as_of: date) -> date:
    try:
        return as_of.replace(year=as_of.year - TRAINING_WINDOW_YEARS)
    except ValueError:
        return as_of.replace(year=as_of.year - TRAINING_WINDOW_YEARS, day=28)


def add_calendar_years(value: date, years: int) -> date:
    try:
        return value.replace(year=value.year + years)
    except ValueError:
        return value.replace(year=value.year + years, day=28)


def add_calendar_months(value: date, months: int) -> date:
    month_index = value.month - 1 + months
    year = value.year + month_index // 12
    month = month_index % 12 + 1
    last_day = calendar.monthrange(year, month)[1]
    return date(year, month, min(value.day, last_day))


def parse_iso_date(raw: Any) -> date | None:
    text = str(raw or "").strip()
    if not text or len(text) < 10 or text[4] != "-":
        return None
    try:
        return date.fromisoformat(text[:10])
    except ValueError:
        return None


def parse_training_hours(raw: Any) -> Decimal | None:
    if raw is None:
        return None
    text = str(raw).strip()
    if not text:
        return None
    normalized = text.replace(",", ".")
    try:
        value = Decimal(normalized)
    except (InvalidOperation, ValueError):
        return None
    if value < 0:
        return None
    return value


def normalize_document_type(raw: Any) -> str | None:
    text = str(raw or "").strip().lower()
    return text or None


def is_active_training_record(record: TrainingSummaryRecord) -> bool:
    status = str(record.lifecycle_status or "active").strip().lower()
    return status == "active"


def is_certificate_record(record: TrainingSummaryRecord) -> bool:
    document_type = normalize_document_type(record.document_type)
    return document_type == TRAINING_DOCUMENT_TYPE_CERTIFICATE


def training_summary_record_from_mapping(raw: Mapping[str, Any]) -> TrainingSummaryRecord:
    metadata = raw.get("metadata")
    document_type = raw.get("document_type")
    if document_type is None and isinstance(metadata, Mapping):
        document_type = metadata.get("document_type")
    return TrainingSummaryRecord(
        title=(str(raw.get("title") or raw.get("course_name") or "").strip() or None),
        completed_at=parse_iso_date(raw.get("completed_at") or raw.get("year_to") or raw.get("year")),
        hours=parse_training_hours(raw.get("hours")),
        document_type=normalize_document_type(document_type),
        lifecycle_status=str(raw.get("lifecycle_status") or "ACTIVE").strip() or "ACTIVE",
    )


def calculate_training_hours_last_5y(
    records: Sequence[TrainingSummaryRecord | Mapping[str, Any]],
    *,
    as_of: date | None = None,
) -> dict[str, Any]:
    ref = as_of or date.today()
    window_start = training_window_start(ref)
    total = Decimal("0")
    included_count = 0

    for raw in records:
        record = raw if isinstance(raw, TrainingSummaryRecord) else training_summary_record_from_mapping(raw)
        if not is_active_training_record(record):
            continue
        completed_at = record.completed_at
        if completed_at is None or completed_at < window_start or completed_at > ref:
            continue
        hours = record.hours
        if hours is None:
            continue
        total += hours
        included_count += 1

    normalized_total = int(total) if total == total.to_integral_value() else float(total)
    return {
        "as_of": ref.isoformat(),
        "window_start": window_start.isoformat(),
        "training_hours_last_5y": normalized_total,
        "qualifying_records_count": included_count,
    }


def calculate_expiring_certificates(
    records: Sequence[TrainingSummaryRecord | Mapping[str, Any]],
    *,
    as_of: date | None = None,
) -> list[ExpiringCertificateSummary]:
    ref = as_of or date.today()
    lookahead_end = add_calendar_months(ref, CERTIFICATE_EXPIRY_LOOKAHEAD_MONTHS)
    results: list[ExpiringCertificateSummary] = []

    for raw in records:
        record = raw if isinstance(raw, TrainingSummaryRecord) else training_summary_record_from_mapping(raw)
        if not is_active_training_record(record):
            continue
        if not is_certificate_record(record):
            continue
        title = str(record.title or "").strip()
        completed_at = record.completed_at
        if not title or completed_at is None:
            continue

        expires_at = add_calendar_years(completed_at, CERTIFICATE_VALIDITY_YEARS)
        if expires_at <= ref or expires_at > lookahead_end:
            continue

        days_remaining = (expires_at - ref).days
        if days_remaining <= 0:
            continue

        results.append(
            ExpiringCertificateSummary(
                title=title,
                expires_at=expires_at,
                days_remaining=days_remaining,
            )
        )

    results.sort(key=lambda item: (item.expires_at, item.title.casefold()))
    return results

File: corpsite-ui/lib/trainingSummary.ts
import {
  formatPersonnelDayDateForDisplay,
  isValidPersonnelDayDateIso,
} from "@/lib/personnelDayDate";
import {
  normalizeIntakeTrainingEntry,
  reconcileTrainingEntryHours,
  resolveIntakeTrainingYearTo,
  type IntakeTrainingEntry,
} from "@/app/intake/_lib/intakeTraining";

export const TRAINING_WINDOW_YEARS = 5;
export const CERTIFICATE_VALIDITY_YEARS = 5;
export const CERTIFICATE_EXPIRY_LOOKAHEAD_MONTHS = 6;
export const TRAINING_DOCUMENT_TYPE_CERTIFICATE = "certificate";

export type TrainingSummaryRecord = {
  title: string | null;
  completedAt: string | null;
  hours: number | null;
  documentType: string | null;
  lifecycleStatus: string | null;
};

export type ExpiringCertificateSummary = {
  title: string;
  expiresAt: string;
  daysRemaining: number;
};

export type TrainingHoursLast5ySummary = {
  asOf: string;
  windowStart: string;
  trainingHoursLast5y: number;
  qualifyingRecordsCount: number;
};

function parseIsoDateOnly(value: string | null | undefined): string | null {
  const trimmed = String(value ?? "").trim();
  if (!trimmed || !isValidPersonnelDayDateIso(trimmed)) return null;
  return trimmed.slice(0, 10);
}

function toUtcDate(isoDate: string): Date {
  const [year, month, day] = isoDate.slice(0, 10).split("-").map(Number);
  return new Date(Date.UTC(year, month - 1, day));
}

function isoFromUtcDate(date: Date): string {
  const year = date.getUTCFullYear();
  const month = String(date.getUTCMonth() + 1).padStart(2, "0");
  const day = String(date.getUTCDate()).padStart(2, "0");
  return `${year}-${month}-${day}`;
}

export function trainingWindowStart(asOfIso: string): string {
  const asOf = toUtcDate(asOfIso);
  const targetYear = asOf.getUTCFullYear() - TRAINING_WINDOW_YEARS;
  const month = asOf.getUTCMonth();
  const day = asOf.getUTCDate();
  const lastDay = new Date(Date.UTC(targetYear, month + 1, 0)).getUTCDate();
  return isoFromUtcDate(new Date(Date.UTC(targetYear, month, Math.min(day, lastDay))));
}

export function addCalendarYears(isoDate: string, years: number): string {
  const date = toUtcDate(isoDate);
  const targetYear = date.getUTCFullYear() + years;
  const month = date.getUTCMonth();
  const day = date.getUTCDate();
  const lastDay = new Date(Date.UTC(targetYear, month + 1, 0)).getUTCDate();
  return isoFromUtcDate(new Date(Date.UTC(targetYear, month, Math.min(day, lastDay))));
}

export function addCalendarMonths(isoDate: string, months: number): string {
  const date = toUtcDate(isoDate);
  const absoluteMonth = date.getUTCMonth() + months;
  const year = date.getUTCFullYear() + Math.floor(absoluteMonth / 12);
  const month = ((absoluteMonth % 12) + 12) % 12;
  const lastDay = new Date(Date.UTC(year, month + 1, 0)).getUTCDate();
  const day = Math.min(date.getUTCDate(), lastDay);
  return isoFromUtcDate(new Date(Date.UTC(year, month, day)));
}

export function parseTrainingHours(value: string | number | null | undefined): number | null {
  if (value === null || value === undefined) return null;
  const text = String(value).trim().replace(",", ".");
  if (!text) return null;
  const parsed = Number(text);
  if (!Number.isFinite(parsed) || parsed < 0) return null;
  return parsed;
}

function normalizeDocumentType(value: string | null | undefined): string | null {
  const text = String(value ?? "").trim().toLowerCase();
  return text || null;
}

export function isActiveTrainingRecord(record: TrainingSummaryRecord): boolean {
  const status = String(record.lifecycleStatus ?? "active").trim().toLowerCase();
  return status === "active";
}

export function isCertificateRecord(record: TrainingSummaryRecord): boolean {
  return normalizeDocumentType(record.documentType) === TRAINING_DOCUMENT_TYPE_CERTIFICATE;
}

export function trainingSummaryRecordFromIntakeEntry(item: IntakeTrainingEntry): TrainingSummaryRecord {
  const normalized = reconcileTrainingEntryHours(normalizeIntakeTrainingEntry(item));
  return {
    title: normalized.course_name.trim() || null,
    completedAt: parseIsoDateOnly(resolveIntakeTrainingYearTo(normalized)),
    hours: parseTrainingHours(normalized.hours),
    documentType: normalizeDocumentType(normalized.document_type),
    lifecycleStatus: "ACTIVE",
  };
}

export type PprTrainingSummarySource = {
  title: string | null;
  completed_at: string | null;
  hours: string | number | null;
  document_type?: string | null;
  lifecycle_status?: string | null;
};

export function trainingSummaryRecordFromPprRecord(record: PprTrainingSummarySource): TrainingSummaryRecord {
  return {
    title: String(record.title ?? "").trim() || null,
    completedAt: parseIsoDateOnly(record.completed_at),
    hours: parseTrainingHours(record.hours),
    documentType: normalizeDocumentType(record.document_type),
    lifecycleStatus: String(record.lifecycle_status ?? "ACTIVE").trim() || "ACTIVE",
  };
}

export function calculateTrainingHoursLast5y(
  records: readonly TrainingSummaryRecord[],
  asOfIso?: string,
): TrainingHoursLast5ySummary {
  const asOf = parseIsoDateOnly(asOfIso ?? isoFromUtcDate(new Date()))!;
  const windowStart = trainingWindowStart(asOf);
  let total = 0;
  let qualifyingRecordsCount = 0;

  for (const record of records) {
    if (!isActiveTrainingRecord(record)) continue;
    const completedAt = parseIsoDateOnly(record.completedAt);
    if (!completedAt || completedAt < windowStart || completedAt > asOf) continue;
    if (record.hours === null) continue;
    total += record.hours;
    qualifyingRecordsCount += 1;
  }

  return {
    asOf,
    windowStart,
    trainingHoursLast5y: total,
    qualifyingRecordsCount,
  };
}

export function calculateExpiringCertificates(
  records: readonly TrainingSummaryRecord[],
  asOfIso?: string,
): ExpiringCertificateSummary[] {
  const asOf = parseIsoDateOnly(asOfIso ?? isoFromUtcDate(new Date()))!;
  const lookaheadEnd = addCalendarMonths(asOf, CERTIFICATE_EXPIRY_LOOKAHEAD_MONTHS);
  const results: ExpiringCertificateSummary[] = [];

  for (const record of records) {
    if (!isActiveTrainingRecord(record)) continue;
    if (!isCertificateRecord(record)) continue;
    const title = String(record.title ?? "").trim();
    const completedAt = parseIsoDateOnly(record.completedAt);
    if (!title || !completedAt) continue;

    const expiresAt = addCalendarYears(completedAt, CERTIFICATE_VALIDITY_YEARS);
    if (expiresAt <= asOf || expiresAt > lookaheadEnd) continue;

    const daysRemaining = Math.floor((toUtcDate(expiresAt).getTime() - toUtcDate(asOf).getTime()) / 86_400_000);
    if (daysRemaining <= 0) continue;

    results.push({ title, expiresAt, daysRemaining });
  }

  return results.sort((left, right) => {
    const byDate = left.expiresAt.localeCompare(right.expiresAt);
    return byDate !== 0 ? byDate : left.title.localeCompare(right.title, "ru");
  });
}

export function formatTrainingSummaryDate(isoDate: string): string {
  return formatPersonnelDayDateForDisplay(isoDate.slice(0, 10), "document") || isoDate;
}
